### ISSUE TITLE
cache.modify INVALIDATE doesn't result in query refetching data

### ISSUE THREAD

**Issue by vedrani** [apollographql/apollo-client#7060]
**Intended outcome:**

My motivation is to invalidate data in cache, so that queries can refetch invalidated data.
(_For that reason I just return `_id` from mutation. I know I can return data with mutation, but this scenario has implication on invalidating data on mutations with side effects._)

For that purpose I'm using `cache.modify` and INVALIDATE flag to invalidate data in cache.

My expectation is that invalidation will result in query refetching data from server because data is not valid any more.

I also tried `evict` and it works, but my impression is that `invalidate` mechanism should allow us to refetch data based on dirty status. Did I miss something in code or did I misunderstood it? Thanks! 

[Codesandbox](https://codesandbox.io/s/dawn-cookies-uxdyb?file=/src/Todo.js:540-577)

**Actual outcome:**

Query that fetches all todos is reobserved but it doesn't refetch data. 

I tried to debug it and it seems cache diff is non existent. After cache.modify invalidation, query is reobserved but data is resolved from cache instead refetched.

What I found is that INVALIDATE is translated into `dirty` status [here](https://github.com/apollographql/apollo-client/blob/main/src/cache/inmemory/entityStore.ts#L429), but Query when calculating diff doesn't see any difference [here](https://github.com/apollographql/apollo-client/blob/main/src/core/QueryManager.ts#L1011-L1013) 

**How to reproduce the issue:**

Open [Codesandbox](https://codesandbox.io/s/dawn-cookies-uxdyb?file=/src/Todo.js:540-577) and try to change status of todo item. 

It doesn't change. 

**Versions**

  System:
    OS: macOS 10.15.5
  Binaries:
    Node: 14.2.0 - ~/.nvm/versions/node/v14.2.0/bin/node
    Yarn: 1.22.5 - ~/.yvm/versions/v1.22.5/bin/yarn
    npm: 6.14.4 - ~/.nvm/versions/node/v14.2.0/bin/npm
  Browsers:
    Chrome: 85.0.4183.121
    Firefox: 75.0
    Safari: 13.1.1
  npmPackages:
    @apollo/client: 3.2.0 => 3.2.0

**Comment by benjamn** [comment]
@vedrani You're right, `INVALIDATE` by itself isn't the best fit with a `cache-first` policy, since the invalidated data remain in the cache, so the next cache read usually succeeds, so no network request happens.

To make the most of `INVALIDATE`, you need to have some control over how queries respond to the invalidation signal. If you're doing a mutation, `refetchQueries` can help, but I am also currently working on a new mutation option that I think can take the place of `refetchQueries` in most cases:
```ts
export default function Todo({ todo }) {
  const [updateTodo] = useMutation(UPDATE_TODO, {
    update(cache, { data: { updateTodo } }) {
      cache.modify({
        id: cache.identify(updateTodo),
        fields(fieldValue, details) {
          return details.INVALIDATE;
        }
      });
    },

    // This callback will be called with each ObservableQuery that was
    // affected by the mutation update function, so you can decide how
    // to refetch the query. The default behavior is the same as calling
    // observableQuery.reobserve(), but calling refetch instead will force
    // a network request.
    reobserveQuery(observableQuery) {
      return observableQuery.refetch();
    },
  });
```
I'll let you know when that's ready for testing (or you can watch #7015 for upcoming betas). Thanks for trying out the `INVALIDATE` feature in the meantime, and for providing such thoughtful feedback!

**Comment by vedrani** [comment]
@benjamn I'm also interested are we going to have utility that will allow us to just watch dirty status in cache for some query? Or query callback will be enough?

What is interesting for me is case of expensive or slow query. Will I have opportunity just to show for example Reload button instead of refetching query immediately?

Thanks!

**Comment by tafelito** [comment]
@benjamn what about queries that are not currently active?

I mean, I have 2 pages, in page 1 I have a list of items, in page 2 I have a function where I can query some data or do something and I need to invalidate data from page 1. Then when I navigate to page 1, I want to show existing data but since cache data is invalidated, I want to refetch data form the server. The fetch policy in page 1 query is cache first

**Comment by stephenhandley** [comment]
@benjamn I've tried invalidating a field using  `"@apollo/client": "3.3.4"` and the following and thread query still appears to be loading from the cache and doesn't reflect the updated field
```
const mutation = useMutation({
  mutation: CreatePost,
  update (cache, {data: {createPost: post}}) {
    const {thread_id} = post;
    cache.modify({
      id: `Thread:${thread_id}`,
      fields: {
        posts (items, {INVALIDATE}) {
          return INVALIDATE;
        }
      }
    });
  },
  reobserveQuery (query) {
    return query.refetch();
  }
});
```

**Comment by justanotherjones** [comment]
> I have to say I've been continually gotcha'd by the over-engineering in Apollo client. The semantics of INVALIDATE should probably be as simple as possible. If I invalidate cache data, no matter what the fetchPolicy is, any subsequent requests to that field (or fields) in the cache should miss and issue a network request.

Definitely agreed. I spent an unfortunate amount of time today misunderstanding INVALIDATE semantics.

@benjamn If INVALIDATE is not meant to trigger a refetch, what is the correct way to indicate to Apollo client that a specific field is invalid and should trigger a refetch?

**Comment by seanemmer** [comment]
I have successfully used the DELETE sentinel instead of the INVALIDATE sentinel to force a refetch with a cache-first fetch policy.  Not sure if this was the new addition mentioned by @benjamn but it appears to work correctly.

**Comment by benjamn** [comment]
PR in progress for the `reobserveQuery` idea I described above: #7827

**Comment by omerman** [comment]
@benjamn I don't understand how `reobserveQuery` is a solution to this problem.. I mean, INVALIDATE indicator should refetch, this is why we all came to this thread, it is intuitive that it would work that way, it is expected to trigger a refetch, no matter what the fetch policy is defined. I think I speak for everyone here when I say that we need a way to mark a query as "needs to refetch next time it is observed", outside react & hooks context, using purely the Apollo client.

**Comment by ab-pm** [comment]
I'm having trouble here deciding between `DELETE` and `INVALIDATE` as well. From the original https://github.com/apollographql/apollo-client/pull/6991, it seems that `INVALIDATE` was always meant to trigger re-observation only while `DELETE` causes a re-fetch due to completely missing data. If not, what else is the difference between them? Do they cause watched queries to return different temporary results while the query is re-run asynchronously?

**Comment by omerman** [comment]
@ab-pm DELETE means that until data comes back, the data is empty, isn't it? and that is not a good user experience..
In my opinion DELETE should only be used when the resource is deleted.. not for refetch..

**Comment by ab-pm** [comment]
@omerman Oops you're right. I hadn't noticed the flicker at first because the backend responded so quickly. I had expected that when a query is refetched, the `QueryResult` would keep its `data` but `loading` would switch to true before the new data arrives. But taking a closer look, it's actually the reverse: `data` becomes `null` when deleting the cache entries, and during the request `loading` stays `false`. That's… puzzling.

**Comment by jbmikk** [comment]
I agree with everyone else here, when I use INVALIDATE I expect Apollo to re-fetch automatically as if the cache was empty.

I expect `cache-first` to return whatever is in the **cache first as long as it is valid content**, and fetch the content stops being valid, but not just the first time. Otherwise what's the point?

If `cache-first` is going work only the first time, then we need a new fetch policy, something like `valid-cache-first`. And this one should be the client's default policy IMHO.

About the `reobserveQuery` it's OK to have something like that if someone wants to be very picky about what happens with each query, but I think we need more reasonable defaults, otherwise we always end up wiring everything as if we were not using a framework.

**Comment by yairopro** [comment]
**So to sum up**

1. The **current workaround** is to use the `DELETE` sentinel instead of `INVALIDATE`.
2. Most of us agree that `INVALIDATE` should invalidate the cache and force a refetch (for `cache-first` policy at least).
3. A new feature `reobserveQuery` is born from this issue 
🐣 Another mouth to feed... I mean, another feature to maintain.

**Comment by dutbu2718** [comment]
Hi everyone,

I'm looking for clarity on best practices regarding marking cache entries as stale without completely evicting data from the cache using `cache.evict().`

Since this issue was originally opened in 2020, I'm wondering what the current best practice is.

Is there a way for Apollo Client to automatically recognize that the cached result was previously invalidated and perform a network refetch when the query becomes active again (component mount), without changing the query's fetchPolicy to `cache-and-network`? 

More specifically, is there a supported API to programmatically determine whether a query result or cache entry has been marked as `INVALIDATE`,  so that the component could do something like:

```ts
if (queryWasInvalidated) {
  query.refetch();
}
```

Today my code need to always use the fetchPolicy `cache-and-network ` when we introduce a mutation.  Could look like as the following: 

```ts
import { ApolloCache } from "@apollo/client";
import { gql } from "@apollo/client";

export const GET_TIMELINE_NOTES_QUERY = gql`
  query GetTimelineNotes($type: String = "timeline") {
    notes(type: $type) {
      id
      title
      content
      type
    }
  }
`;

export const GET_TIMELINE_DETAILS_QUERY = gql`
  query GetTimelineDetails($id: ID!, $type: String = "timeline") {
    note(id: $id, type: $type) {
      id
      title
      content
      type
      metadata {
        eventTime
        location
      }
    }
  }
`;

export const GET_NOTES_QUERY = gql`
  query GetNotes($type: String = "standard") {
    notes(type: $type) {
      id
      content
      type
    }
  }
`;

export const UPDATE_NOTE_MUTATION = gql`
  mutation UpdateNote($id: ID!, $text: String!) {
    updateNote(id: $id, text: $text) {
      id
      text
    }
  }
`;

/**
 * @description
 * Used to refetch all `active` timeline note queries.
 */
export const REFETCH_ALL_TIMELINE_NOTE_QUERIES = [
  GET_TIMELINE_NOTES_QUERY,
  GET_TIMELINE_DETAILS_QUERY,
];

/**
 * @description
 * Mark all `active` and `inactive` timeline note queries as stale.
 *
 * @remarks
 * We intentionally do not modify the cache here because we do not know how
 * `apollo.watchQuery` instances consume this data (for example, how their fetch
 * policies or query documents work). Updating the cache directly
 * (such as with `cache.evict`) could therefore introduce unintended side effects.
 */
export function invalidateAllTimelineNotes(cache: ApolloCache) {
  cache.modify({
    id: "ROOT_QUERY",
    fields: {
      notes(value, details) {
        const isTimelineEntry =
          details.storeFieldName.includes(`"type":timeline`);

        if (isTimelineEntry) {
          return details.INVALIDATE;
        }

        return value;
      },
    },
  });
}

function updateStandardNote(noteId: string, text: string) {
  return apollo.mutate({
    mutation: UPDATE_NOTE_MUTATION,
    variables: {
      id: noteId,
      text: text,
    },
    errorPolicy: "all",
    refetchQueries: [GET_NOTES_QUERY, ...REFETCH_ALL_TIMELINE_NOTE_QUERIES],
    update: (cache) => {
      invalidateAllTimelineNotes(cache);
    },
  });
}

function getStandardNotes() {
  return apollo.watchQuery({
    query: GET_NOTES_QUERY,
    fetchPolicy: "cache-and-network",
    errorPolicy: "all",
  });
}

function getTimelineNotes() {
  return apollo.watchQuery({
    query: GET_TIMELINE_NOTES_QUERY,
    fetchPolicy: "cache-and-network", // Use to always fetch as INVALIDATE does not make cache-only fetch after mutation.
    errorPolicy: "all",
  });
}

```

**Note:** We are aware that a mutation could return the data required by both GET_TIMELINE_NOTES_QUERY and GET_TIMELINE_DETAILS_QUERY as part of its response. However, this approach could introduce unnecessary coupling between bounded contexts and allow concerns from one context to leak into another.

**Comment by jerelmiller** [comment]
Hey @dutbu2718 👋 

tl;dr; no. `evict` (or `cache.modify` with `DELETE`) is still your best bet for this job so nothing has really changed in this regard. 

This is something I've wanted/thought about and want to include as part of v5 (see the last half of [this comment](https://github.com/apollographql/apollo-feature-requests/issues/419#issuecomment-2463021311) where I talk about almost the same ask you have 🙂).

Out of curiosity, assuming we had something where `INVALIDATE` marked a field as stale, what would you expect `cache-first` to return when that cache data is stale?

**Comment by mogelbrod** [comment]
My organization is very much interested in a proper way to mark specific field nodes as stale. We'd happily help test any draft APIs that come up. Right now we have to work around this with a custom `queryTTL()` / `useQueryTTL()` API that tracks time since last query per document and arguments, which is quite ugly.

> Out of curiosity, assuming we had something where `INVALIDATE` marked a field as stale, what would you expect `cache-first` to return when that cache data is stale?

I'd assume `apollo.query()` with `cache-first` would trigger a request and only resolve once done. `useQuery()` with `cache-first` would on the other hand render the stale data while simultaneously triggering a request.

**Comment by ab-pm** [comment]
> This is something I want to include as part of v5

Great news! I'm not sure how much the internal data structures will need to change represent stored-but-stale values in the cache, but maybe also consider a time-to-live (TTL) timestamp (as suggested e.g. back in #621), i.e. a point in time from when the value will be considered stale. I'd love to be able to set a TTL policy per field (or in each query that fetches the field). This could be implemented manually I guess once we get `INVALIDATE`, but it would be a really great feature so maybe already prepare for it later.

**Comment by jerelmiller** [comment]
Cache TTL is definitely something else on my mind and something I think we need to add. I know it would solve a lot of pain for some 🙂. Ideally we can get that in v4 because I think we can at least do cache TTL without breaking changes.

> `useQuery()` with `cache-first` would on the other hand render the stale data while simultaneously triggering a request.

This is part of the reason I asked the question above. What you're describing is essentially `cache-and-network` behavior which is meant as a render any (potentially) stale cache data while making a network request. `cache-first` traditionally is an all or nothing, so if you get a full cache result, you won't see a network request. The question is whether "staleness" is treated the same as a cache miss (in which case `cache-first` would return `data: undefined`, not the stale data).

This is where I struggle to figure out whether this feature can belong in v4 or whether it would need to wait until v5. `cache-first` that can return a cache result and make a network request might be considered a breaking change (or at least introduce confusion). But glad to know there is interest in this! And helps me validate that my thinking is on the right track 😆

**Comment by ab-pm** [comment]
Btw I agree with @mogelbrod's expectations.

> What you're describing is essentially `cache-and-network` behavior which is meant as a render any (potentially) stale cache data while making a network request

I don't quite follow. `cache-and-network` *forces* the network request regardless whether data is in the cache, no? `cache-first` should make a network request only if it didn't find valid data in the cache (whether missing or stale). Both should return possibly-stale data if already available as soon as possible.

I doubt people would consider it a breaking change for `cache-first` to return data while also making a request. Why should it be either-or? It's "both" over time anyway if a query needs network at first but can later be fulfilled from cache. There's a lot of confusion already what queries return while the network request is active, see e.g. #7605. I think this at best just needs better documentation (and possibly more result properties to distinguish the states).

**Comment by dutbu2718** [comment]
Hi @jerelmiller! 👋👋

> 
> > `useQuery()` with `cache-first` would on the other hand render the stale data while simultaneously triggering a request.
> 
> This is part of the reason I asked the question above. What you're describing is essentially `cache-and-network` behavior which is meant as a render any (potentially) stale cache data while making a network request. `cache-first` traditionally is an all or nothing, so if you get a full cache result, you won't see a network request. The question is whether "staleness" is treated the same as a cache miss (in which case `cache-first` would return `data: undefined`, not the stale data).
> 

Sorry if I misinterpreted the other comments in this thread (I have worked with React, Vue and currently with Angular 22). Shouldn’t  `useQuery()`  (I assume React) and `apollo.query()` (I assume Angular and other frameworks) have the same behavior independent of the framework? 

**I'd assume  `cache-first `on ANY frameworks **should mean only resolve once done**.** Or at least have the same behavior independent of what we have decided of course. 😄 

However, I think the current `useQuery()/apollo.query()` API implementation is missing an “onStale” callback like Relay`s useSubscribeToInvalidationState hook (https://relay.dev/docs/guided-tour/reusing-cached-data/staleness-of-data/#subscribing-to-data-invalidation). 

This would allow the consumers to handle whatever “staleness” means for them.

```ts
useQuery(GET_MY_QUERY, {
  fetchPolicy: “cache-first“,
  onStale: (context) => {
    // “Staleness” can be ambiguous.
    // - Does it mean GET_MY_QUERY TTL has expired?
    // - Has someone invalidated the GET_MY_QUERY in the cache?

    // Based on the above I may want to handle it differently after component mount.

    // Imperatively handle “staleness”
    context.refetch();
    // Or on a infinite query use case
    context.fetchMore();
   // Or on a optimistic update use case. I need to maybe do things in my components state first. Like scroll-restore 
       ...
  },

  // If we are really crazy and for the offline first people like myself ;)
  onCacheMiss: () => {
    // Load from SQLite or DuckDB Wasm
    // As a have an offline first application
  }}
```

**Comment by ab-pm** [comment]
@dutbu2718 

> only resolve once done

Apollo client's `.query()` method returns an [`ObservableQuery`](https://www.apollographql.com/docs/react/api/core/ObservableQuery) that will provide multiple results over time, not a `Promise` that fulfills (resolves) only once.

But yes, an `onInvalidated` (`onStale`) callback would be a good idea as well. This would allow customising the invalidation behaviour, though I think the default (in an active query) should be refetching - this is what the OP of this issue is complaining about. This change of behaviour - if considered a breaking change - could also be made opt-in with such a policy setting (which probably should be independent of the `fetchPolicy`?).

**Comment by dutbu2718** [comment]
@ab-pm 

> But yes, an `onInvalidated` (`onStale`) callback would be a good idea as well. This would allow customising the invalidation behaviour, though I think the default (in an active query) should be refetching - this is what the OP of this issue is complaining about. This change of behaviour - if considered a breaking change - could also be made opt-in with such a policy setting (which probably should be independent of the `fetchPolicy`?).

I whole agree with Andreas. For our organization this new behaviour would not be a huge deal for us. We would really love it.

But from my not so fun experience from migrating from Apollo Client v3 to v4 on multiple repos. We should let other organizations have the breathing room to gradually opt-in to this new default behaviour.
